pkg/paths/path.go

```diff
@@ -0,0 +1,147 @@
+}
+
+// Valid returns if the unencrypted path is valid, which is different from being empty.
+func (path Unencrypted) Valid() bool {
+	return path.valid
+}
+
+// Raw returns the original raw path for the Unencrypted.
+func (path Unencrypted) Raw() string {
+	return path.raw
+}
+
+// String returns a human readable form of the Unencrypted.
+func (path Unencrypted) String() string {
+	if !path.valid {
+		return "<unencrypted-invalid-path>"
+	}
+	return path.Raw()
+}
+
+// Consume attempts to remove the prefix from the Unencrypted, and reports true
+// if it was successful.
+func (path Unencrypted) Consume(prefix Unencrypted) (Unencrypted, bool) {
+	if len(path.raw) >= len(prefix.raw) && path.raw[:len(prefix.raw)] == prefix.raw {
+		return NewUnencrypted(path.raw[len(prefix.raw):]), true
+	}
+	return path, false
+}
+
+// Iterator returns an iterator over the components of the Unencrypted.
+func (path Unencrypted) Iterator() Iterator {
+	return Iterator{raw: path.raw}
+}
+
+//
+// encrypted path
+//
+
+// NewEncrypted takes a raw encrypted path and returns it wrapped.
+func NewEncrypted(raw string) Encrypted {
+	return Encrypted{raw: raw, valid: true}
+}
+
+// Valid returns if the encrypted path is valid, which is different from being empty.
+func (path Encrypted) Valid() bool {
+	return path.valid
+}
+
+// Raw returns the original path for the Encrypted.
+func (path Encrypted) Raw() string {
+	return path.raw
+}
+
+// String returns a human readable form of the Encrypted.
+func (path Encrypted) String() string {
+	if !path.valid {
+		return "<encrypted-invalid-path>"
+	}
+	return path.Raw()
+}
+
+// Consume attempts to remove the prefix from the Encrypted, and reports true
+// if it was successful.
+func (path Encrypted) Consume(prefix Encrypted) (Encrypted, bool) {
+	if len(path.raw) >= len(prefix.raw) && path.raw[:len(prefix.raw)] == prefix.raw {
+		return NewEncrypted(path.raw[len(prefix.raw):]), true
+	}
+	return path, false
+}
+
+// Iterator returns an iterator over the components of the Encrypted.
+func (path Encrypted) Iterator() Iterator {
+	return Iterator{raw: path.raw}
+}
+
+//
+// path component iteration
+//
+
+// Iterator allows one to efficiently iterate over components of a path.
+type Iterator struct {
+	raw       string
+	consumed  int
+	lastEmpty bool
+}
+
+// NewIterator returns an Iterator for components of the provided raw path.
+func NewIterator(raw string) Iterator {
+	return Iterator{raw: raw}
+}
+
+// Consumed reports how much of the path has been consumed.
+func (pi Iterator) Consumed() string { return pi.raw[:pi.consumed] }
+
+// Remaining reports how much of the path is remaining.
+func (pi Iterator) Remaining() string { return pi.raw[pi.consumed:] }
+
+// Done reports if the path has been fully consumed.
+func (pi Iterator) Done() bool { return len(pi.raw) == pi.consumed && !pi.lastEmpty }
+
+// Next returns the first component of the path, consuming it.
+func (pi *Iterator) Next() string {
```

Comment:
also https://storjlabs.atlassian.net/browse/V3-1787

Reply:
A "path" in this package never includes a bucket name, so I think it's encoding agnostic (as long as the encoding doesn't allow '/'). Can you explain more what impact V3-1787 has?